client/web/src/repo/blob/codemirror/token-selection/code-intel-tooltips.ts

```diff
@@ -1,105 +1,129 @@
+export const setFocusedOccurrenceTooltip = StateEffect.define<Tooltip | null>()
+const setPinnedCodeIntelTooltipState = StateEffect.define<CodeIntelTooltipState>()
+const setHoveredCodeIntelTooltipState = StateEffect.define<CodeIntelTooltipState>()
+
+/**
+ * {@link StateField} storing focused (selected), hovered and pinned {@link Occurrence}s and {@link Tooltip}s associate with them.
+ */
+export const codeIntelTooltipsState = StateField.define<Record<CodeIntelTooltipTrigger, CodeIntelTooltipState>>({
+    create() {
+        return { hover: null, focus: null, pin: null }
     },
-    provide: field => showTooltip.from(field),
-})
-export const setHoveredOccurrenceEffect = StateEffect.define<Occurrence | null>()
-export const hoveredOccurrenceField = StateField.define<Occurrence | null>({
-    create: () => null,
     update(value, transaction) {
         for (const effect of transaction.effects) {
-            if (effect.is(setHoveredOccurrenceEffect)) {
-                value = effect.value
+            if (effect.is(setFocusedOccurrence)) {
+                return {
+                    ...value,
+                    focus: effect.value ? { occurrence: effect.value, tooltip: null } : null,
+                }
+            }
+            if (effect.is(setFocusedOccurrenceTooltip)) {
+                return {
+                    ...value,
+                    focus: value.focus?.occurrence ? { ...value.focus, tooltip: effect.value } : null,
+                }
+            }
+
+            if (effect.is(setHoveredCodeIntelTooltipState)) {
+                return { ...value, hover: effect.value }
+            }
+
+            if (effect.is(setPinnedCodeIntelTooltipState)) {
+                return { ...value, pin: effect.value }
             }
         }
+
         return value
     },
+    provide(field) {
+        return [
+            showTooltip.computeN([field], state => Object.values(state.field(field)).map(val => val?.tooltip ?? null)),
+        ]
+    },
 })
 
-const getPinnedOccurrence = (view: EditorView, pin: LineOrPositionOrRange | null): Occurrence | null => {
-    if (!pin || !pin.line || !pin.character) {
-        return null
+export const getCodeIntelTooltipState = (
+    view: EditorView,
+    key: CodeIntelTooltipTrigger
+): { occurrence: Occurrence; tooltip: Tooltip | null } | null => view.state.field(codeIntelTooltipsState)[key]
+
+const focusOccurrence = (view: EditorView, occurrence: Occurrence): void => {
+    const offset = positionToOffset(view.state.doc, occurrence.range.end)
+    if (offset !== null) {
+        const node = view.domAtPos(offset).node
+        const element = node instanceof HTMLElement ? node : node.parentElement
+        const lineEl = element?.matches('.cm-line') ? element : element?.closest('.cm-line')
+        const interactiveOccurrenceEl = lineEl?.querySelector<HTMLElement>('.interactive-occurrence')
+        if (interactiveOccurrenceEl) {
+            interactiveOccurrenceEl?.focus()
```

Comment:
```suggestion
            interactiveOccurrenceEl.focus()
```

You're already checking for nullability in the line before 😅